### DISCUSSION TITLE
Add punctuation in raw textual data

### DISCUSSION THREAD

**Discussion by kamrapooja** [explosion/spaCy#10420] [Help: Other Questions] [answered]
I have raw data with me without punctuation mark. Is there any way to add punctuation in the data.
e.g. "welcome back in this week you learn how to use the karasz layer in tensorflow to implement word embeddings which is one of the most important ideas in natural language processing if you have a vocabulary size of say ten thousand words rather than using the numbers from one to ten thousand to represent these words can you have a better way to represent those numbers so this week you'll learn about word embeddings and also get to train your own neural network for text classification and it's really interesting that i really liked how embeddings work for representing the semantics of the word so now instead of the word just being a number it's like a vector in north dimensional space we're using sixteen and thirty two dimensions in different examples so for

Here is no end or start of sentence in data. 
Is there any way to add punctuation using spicy?

**Comment by polm** [comment, marked as answer]
spaCy doesn't have any kind of model that would add punctuation. In general, text generation / NLG is out of scope for spaCy. 

You could use the parser / senter to get sentence boundaries pretty easily. You might be able to do something with the parse tree, which I think might work on unpunctuated text, but it's probably better and less work to use a specialized model.

In general I would expect this to modelled as a seq2seq problem. It looks like in academic literature this problem is sometimes referred to as "punctuation restoration".

**Comment by kamrapooja** [comment]
Ok. Thanks for Response

**Comment by richieM** [comment]
@kamrapooja If you still need punctuation on text, check out something like https://huggingface.co/felflare/bert-restore-punctuation
